File: 2342-max-sum-of-a-pair-with-equal-sum-of-digits/2342-max-sum-of-a-pair-with-equal-sum-of-digits.cs
public class Solution {
    public int MaximumSum(int[] nums) {
        var dic = new Dictionary<int, int>();
        int ret = int.MinValue;
        
        foreach(int n in nums){
            var digit = GetSumOfDigits(n);
            if(dic.ContainsKey(digit)) 
            {
                ret = Math.Max(ret, dic[digit]+n);
                if(dic[digit] >= n) continue;
            }
            dic[digit] = n;
        }
        return (ret == int.MinValue) ? -1 : ret;
        
    }
    public int GetSumOfDigits(int num){
        int ret = 0;
        while(num > 0)
        {
            ret+=num%10;
            num /= 10;
        }
        return ret;
    }
}

File: 2342-max-sum-of-a-pair-with-equal-sum-of-digits/2342-max-sum-of-a-pair-with-equal-sum-of-digits.ts
function maximumSum(nums: number[]): number {
    const map: Map<number, number> = new Map();
    let ret = -1;
    
    const digitSum = (num) => [...num.toString()].reduce((acc, val) => acc + parseInt(val), 0);
    
    for(let n of nums){
        const digit = digitSum(n);
        if(map.has(digit)){
            ret = Math.max(ret, map.get(digit)+n);
            if(map.get(digit) >= n) continue;
        }
        map.set(digit, n);
    }
    
    return ret;
};